macros/scripts/maps.asm

```diff
@@ -1,3 +1,12 @@
+def_objects: MACRO
+	IF DEF(_NUM_OBJECTS)
+		PURGE _NUM_OBJECTS
+	ENDC
+_NUM_OBJECTS EQUS "_NUM_OBJECTS_\@"
+	db _NUM_OBJECTS
```

Comment:
How/when does this get resolved?

Reply:
`db _NUM_OBJECTS_#` occurs before `_NUM_OBJECTS_#` gets assigned a value, so instead of the `db` being whatever value `_NUM_OBJECTS_#` had at the time, it takes the final value of `_NUM_OBJECTS_#`, i.e. after all the `object` declarations have finished incrementing it. The `\@` serves the same purpose here as in https://github.com/pret/pokered/blob/master/home/text_script.asm#L69-L75 : it creates a unique symbol for each use of `def_objects`, so each map gets its own count.

Previous discussion of this approach in pokecrystal: https://github.com/pret/pokecrystal/pull/754

Reply:
I guess what I'm not understanding is how `_NUM_OBJECTS_#` ever gets incremented.
I can see `_NUM_OBJECTS` is getting incremented, but how is that final value making its way back into `_NUM_OBJECTS_#`?

Reply:
In `_NUM_OBJECTS = _NUM_OBJECTS + 1`, first `_NUM_OBJECTS` gets evaluated to `_NUM_OBJECTS_#`, and then `_NUM_OBJECTS_# = _NUM_OBJECTS_# + 1` gets evaluated, incrementing `_NUM_OBJECTS_#`. (It's a documented feature of rgbasm, not some weird hack: "An `EQUS` can be expanded to a string that contains another `EQUS` and it will be expanded as well.")

Reply:
There's a more apparent example of this feature within the `warp` macro:

```
_WARP_TO_NAME EQUS "_WARP_TO_NUM_{d:{_NUM_WARPS}}"
_WARP_TO_NAME EQUS "warp_to \1, \2, _WARP_TO_WIDTH"
    PURGE _WARP_TO_NAME
```

The first `_WARP_TO_NAME EQUS ...` line defines the `_WARP_TO_NAME` constant since `_WARP_TO_NAME` is not yet defined; the second defines a constant named the value of  `_WARP_TO_NAME`, since now it has the value `_WARP_TO_NUM_#`. (And then `_WARP_TO_NAME` is immediately `PURGE`d so the next `warp` can redefine `_WARP_TO_NAME` as the next `_WARP_TO_NUM_#`.)

Reply:
The `_WARP_TO_NAME` example makes sense. My confusion with `_NUM_OBJECTS` has to do with this part when you said:
> first `_NUM_OBJECTS` gets evaluated to `_NUM_OBJECTS_#`

but `_NUM_OBJECTS` immediately gets overwritten to be exactly `0` so I'm not sure how it continues to get evaluated to `_NUM_OBJECTS_#`

Reply:
Oh I understand it now. And I see how your `_WARP_TO_NAME` example applies equally here.
That sure seems counterintuitive, but the simplified map data is definitely handy.

Reply:
Perhaps some comments like this could help others avoid my same confusion (since I was assuming the left side of `EQUS` would not be expanded):
```
_NUM_OBJECTS = 0 ; _NUM_OBJECTS_\@ = 0
```

```
_NUM_OBJECTS = _NUM_OBJECTS + 1 ; _NUM_OBJECTS_\@ = _NUM_OBJECTS_\@ + 1